### ISSUE TITLE
Spack does not bootstrap with external compilers

### ISSUE THREAD

**Issue by adityakavalur** [LLNL/spack#28508]
### Steps to reproduce

Bootstrapping with external compilers that need `LD_LIBRARY_PATH` is currently not possible as the process does not use `compilers.yaml`.  A workaround is to always run bootstrap before adding external compilers.

Additionally, maybe this needs to be documented on the "Getting Started" page in the "Bootstrapping Clingo" section, while the above is being addressed?

### Error message

```
[kavalurav@scfocus18 spack]$ spack install openkim-models
==> Bootstrapping clingo from pre-built binaries
==> Fetching https://mirror.spack.io/bootstrap/github-actions/v0.1/build_cache/linux-rhel5-x86_64/gcc-9.3.0/clingo-bootstrap-spack/linux-rhel5-x86_64-gcc-9.3.0-clingo-bootstrap-spack-hmnv6gk5wha64k6r3s7hid35mzvhkuot.spack
==> Installing "clingo-bootstrap@spack%gcc@9.3.0~docs~ipo+python build_type=Release arch=linux-rhel5-x86_64" from a buildcache
==> Bootstrapping patchelf from pre-built binaries
==> Installing patchelf-0.13.1-q2phgoulbx4m4kbctzlyptsczbw6e6zi
==> No binary for patchelf-0.13.1-q2phgoulbx4m4kbctzlyptsczbw6e6zi found: installing from source
==> Fetching https://github.com/NixOS/patchelf/releases/download/0.13.1/patchelf-0.13.1.tar.gz
==> No patches needed for patchelf
==> patchelf: Executing phase: 'autoreconf'
==> patchelf: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 77:
    '/tmp/kavalurav/spack-stage/spack-stage-patchelf-0.13.1-q2phgoulbx4m4kbctzlyptsczbw6e6zi/spack-src/configure' '--prefix=/home/kavalurav/.spack/bootstrap/store/linux-centos7-x86_64/gcc-8.2.0/patchelf-0.13.1-q2phgoulbx4m4kbctzlyptsczbw6e6zi'

2 errors found in build log:
     7     checking for gawk... gawk
     8     checking whether make sets $(MAKE)... yes
     9     checking whether make supports nested variables... yes
     10    checking whether make supports the include directive... yes (GNU style)
     11    checking for gcc... /lstr/home/kavalurav/spack/lib/spack/env/gcc/gcc
     12    checking whether the C compiler works... no
  >> 13    configure: error: in `/tmp/kavalurav/spack-stage/spack-stage-patchelf-0.13.1-q2phgoulbx4m4kbctzlyptsczbw6e6zi/spack-src':
  >> 14    configure: error: C compiler cannot create executables
     15    See `config.log' for more details

See build log for details:
  /tmp/kavalurav/spack-stage/spack-stage-patchelf-0.13.1-q2phgoulbx4m4kbctzlyptsczbw6e6zi/spack-build-out.txt

==> Bootstrapping clingo from sources
[+] /usr (external bison-3.0.4-6by6mwnentffr2smzgpc4r6w4trgki7t)
==> Installing pkgconf-1.8.0-4qrvznh6qmp75hhnhjp37atgcckjwnf2
==> No binary for pkgconf-1.8.0-4qrvznh6qmp75hhnhjp37atgcckjwnf2 found: installing from source
==> Using cached archive: /lstr/home/kavalurav/spack/var/spack/cache/_source-cache/archive/ef/ef9c7e61822b7cb8356e6e9e1dca58d9556f3200d78acab35e4347e9d4c2bbaf.tar.xz
==> No patches needed for pkgconf
==> pkgconf: Executing phase: 'autoreconf'
==> pkgconf: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 77:
    '/tmp/kavalurav/spack-stage/spack-stage-pkgconf-1.8.0-4qrvznh6qmp75hhnhjp37atgcckjwnf2/spack-src/configure' '--prefix=/home/kavalurav/.spack/bootstrap/store/linux-centos7-x86_64/gcc-8.2.0/pkgconf-1.8.0-4qrvznh6qmp75hhnhjp37atgcckjwnf2'

2 errors found in build log:
  >> 6    configure: error: in `/tmp/kavalurav/spack-stage/spack-stage-pkgconf-1.8.0-4qrvznh6qmp75hhnhjp37atgcckjwnf2/spack-src':
  >> 7    configure: error: C compiler cannot create executables
     8    See `config.log' for more details

See build log for details:
  /tmp/kavalurav/spack-stage/spack-stage-pkgconf-1.8.0-4qrvznh6qmp75hhnhjp37atgcckjwnf2/spack-build-out.txt

==> Warning: Skipping build of ncurses-6.2-nbig43rq3uugefzgrstfaz4wymjywiol since pkgconf-1.8.0-4qrvznh6qmp75hhnhjp37atgcckjwnf2 failed
==> Warning: Skipping build of cmake-3.22.1-zr4siq4tqsnvxwj2toke2p4ezsboe745 since ncurses-6.2-nbig43rq3uugefzgrstfaz4wymjywiol failed
==> Warning: Skipping build of clingo-bootstrap-spack-mq6yfzoekt4xhmb6cidbpjmqwahw3jis since cmake-3.22.1-zr4siq4tqsnvxwj2toke2p4ezsboe745 failed
==> Warning: Skipping build of readline-8.1-iepsygqsde5da5zrdk2hu4alace672nx since ncurses-6.2-nbig43rq3uugefzgrstfaz4wymjywiol failed
==> Warning: Skipping build of gdbm-1.19-7e2k3j4u6irr7yk6xdpaxidzlhoq7scd since readline-8.1-iepsygqsde5da5zrdk2hu4alace672nx failed
==> Warning: Skipping build of perl-5.34.0-2qjbggi23gvvqyrzcwva26abl44ez6pq since gdbm-1.19-7e2k3j4u6irr7yk6xdpaxidzlhoq7scd failed
==> Warning: Skipping build of openssl-1.1.1m-5twjb37nkt2255wa7evgbdvoqloat6uq since perl-5.34.0-2qjbggi23gvvqyrzcwva26abl44ez6pq failed
==> Error: cannot bootstrap the "clingo" Python module from spec "clingo-bootstrap@spack+python %gcc target=x86_64" due to the following failures:
    'github-actions' raised RuntimeError: cannot bootstrap any of the patchelf executables from spec "patchelf@0.13.1:0.13.99 %gcc target=x86_64"
    'spack-install' raised InstallError: Terminating after first install failure: ProcessError: Command exited with status 77:
    '/tmp/kavalurav/spack-stage/spack-stage-pkgconf-1.8.0-4qrvznh6qmp75hhnhjp37atgcckjwnf2/spack-src/configure' '--prefix=/home/kavalurav/.spack/bootstrap/store/linux-centos7-x86_64/gcc-8.2.0/pkgconf-1.8.0-4qrvznh6qmp75hhnhjp37atgcckjwnf2'
    Please run `spack -d spec zlib` for more verbose error messages
```

with the below compilers.yaml file
```
[kavalurav@scfocus18 spack]$ cat /home/kavalurav/.spack/bootstrap/config/linux/compilers.yaml 
compilers:
- compiler:
    spec: gcc@4.8.5
    paths:
      cc: /bin/gcc
      cxx: /bin/g++
      f77: /bin/gfortran
      fc: /bin/gfortran
    flags: {}
    operating_system: centos7
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
- compiler:
    spec: gcc@8.2.0
    paths:
      cc: /usr/nic/compiler/gcc/8.2/bin/gcc
      cxx: /usr/nic/compiler/gcc/8.2/bin/g++
      f77: /usr/nic/compiler/gcc/8.2/bin/gfortran
      fc: /usr/nic/compiler/gcc/8.2/bin/gfortran
    flags: {}
    operating_system: centos7
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
```

### Information on your system

```
[kavalurav@scfocus18 ~]$ spack debug report
* **Spack:** 0.17.1-915-c11ce3b
* **Python:** 3.9.2
* **Platform:** linux-centos7-skylake_avx512
* **Concretizer:** clingo
```

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by adityakavalur** [comment]
@alalazo Thanks for the help! much appreciated! 
I made this to document the issue and maybe a modification to the docs? Feel free to close, however.

**Comment by alalazo** [comment]
Solved as of v0.18 when we started bootstrapping patchelf from binaries too.
